### PR TITLE
Remove the postings highlighter and make unified the default highlighter choice

### PR DESCRIPTION
**Pull request by jimczi**
This change removes the `postings` highlighter. This highlighter has been removed from Lucene master (7.x) because it behaves
exactly like the `unified` highlighter when index_options is set to `offsets`:
https://issues.apache.org/jira/browse/LUCENE-7815

It also makes the `unified` highlighter the default choice for highlighting a field (if `type` is not provided).
The strategy used internally by this highlighter remain the same as before, it checks `term_vectors` first, then `postings` and ultimately it re-analyzes the text.
This change also rewrites the docs so that the options that the `unified` highlighter cannot handle are clearly marked as such.
There are few features that the `unified` highlighter is not able to handle which is why the other highlighters (`plain` and `fvh`) are still available.
I'll open separate issues for these features and we'll deprecate the `fvh` and `plain` highlighters when full support for these features have been added to the `unified`.

### CODE REVIEW

File: docs/reference/migration/migrate_6_0/search.asciidoc [hunk 1]
```diff
@@ -98,3 +98,14 @@ but the only reason why it has not been deprecated too is because it is used
 for the `random_score` function. If you really need access to the id of
 documents for sorting, aggregations or search scripts, the recommandation is
 to duplicate the id as a field in the document.
+
+==== Highlighers
```
Comment (clintongormley) [inline docs/reference/migration/migrate_6_0/search.asciidoc]: missing t

File: docs/reference/migration/migrate_6_0/search.asciidoc [hunk 2]
```diff
@@ -98,3 +98,14 @@ but the only reason why it has not been deprecated too is because it is used
...
+
+The `unified` highlighter is the new default choice for highlighter.
+The offset strategy for each field is picked internally by this highlighter depending on the
```
Comment (clintongormley) [inline docs/reference/migration/migrate_6_0/search.asciidoc]: What do you mean by the offset strategy?

File: docs/reference/search/request/highlighting.asciidoc [hunk 3]
```diff
@@ -1,9 +1,7 @@
 [[search-request-highlighting]]
 === Highlighting
 
-Allows to highlight search results on one or more fields. The
-implementation uses either the lucene `plain` highlighter, the
-fast vector highlighter (`fvh`) or `postings` highlighter.
+Allows to highlight search results on one or more fields.
```
Comment (clintongormley) [inline docs/reference/search/request/highlighting.asciidoc]: "Highlighters allow you to produce highlighted snippets from one or more fields in your search results."

File: docs/reference/search/request/highlighting.asciidoc [hunk 4]
```diff
@@ -45,35 +43,43 @@ from versions before 5.0) that match the expression to be highlighted.
...
 highlight on a field anyway, you have to provide the field name explicitly.
 
-[[plain-highlighter]]
-==== Plain highlighter
+[[unified-highlighter]]
+==== Unified Highlighter
 
-The default choice of highlighter is of type `plain` and uses the Lucene highlighter.
-It tries hard to reflect the query matching logic in terms of understanding word importance and any word positioning criteria in phrase queries.
+The default choice of highlighter is of type `unified` and uses the Lucene Unified highlighter.
```
Comment (clintongormley) [inline docs/reference/search/request/highlighting.asciidoc]: The `unified` highlighter (which is used by default if no highlighter `type` is specified) uses the Lucene Unified Highlighter.

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 5]
```diff
@@ -279,195 +276,190 @@ public void testEnsureNoNegativeOffsets() throws Exception {
...
-                        .endObject()
-                        .startObject("attachments").startObject("properties")
-                            .startObject("body")
-                            .field("type", "text")
-                            .field("store", false)
-                            .field("term_vector", "no")
-                        .endObject().endObject().endObject()
-                        .endObject().endObject().endObject()));
+            .addMapping("type1", jsonBuilder().startObject().startObject("type1").startObject("properties")
+                // we don't store title and don't use term vector, now lets see if it works...
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Can you keep the indentation?

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java [hunk 6]
```diff
@@ -279,195 +276,190 @@ public void testEnsureNoNegativeOffsets() throws Exception {
...
-                                    .field("type", "text")
-                                    .field("store", false)
-                                    .field("term_vector", "with_positions_offsets")
-                                .endObject()
-                            .endObject()
-                        .endObject()
-                        .endObject().endObject().endObject()));
+            .addMapping("type1", jsonBuilder().startObject().startObject("type1").startObject("properties")
+                // we don't store title, now lets see if it works...
+                .startObject("title")
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Here too.

File: docs/reference/search/request/highlighting.asciidoc [hunk 7]
```diff
@@ -45,35 +43,43 @@ from versions before 5.0) that match the expression to be highlighted.
...
-If you want to highlight a lot of fields in a lot of documents with complex queries this highlighter will not be fast.
-In its efforts to accurately reflect query logic it creates a tiny in-memory index and re-runs the original query criteria through
-Lucene's query execution planner to get access to low-level match information on the current document.
-This is repeated for every field and every document that needs highlighting. If this presents a performance issue in your system consider using an alternative highlighter.
+This highlighter breaks the text into sentences and scores individual sentences as
+if they were documents in this corpus, using the BM25 algorithm.
+It also supports accurate phrase and multi-term (fuzzy, prefix, regex) highlighting.
+
+[float]
+===== Offsets Strategy
```
Comment (clintongormley) [inline docs/reference/search/request/highlighting.asciidoc]: Offsets Strategy requires some explanation, perhaps:

In order to create meaningful search snippets from the terms being queried, a highlighter needs to know the start and end character offsets of each word in the original text.  These offsets can be obtained from:

* The postings list (fields mapped as `"index_options": "offsets"`).
* Term vectors (fields mapped as `"term_vectors": "with_positions_offsets"`).
* The original field, by reanalysing the text on-the-fly.

**Review by clintongormley — COMMENTED** [review]
Added a few docs comments

File: docs/reference/search/request/highlighting.asciidoc [hunk 8]
```diff
@@ -45,35 +43,43 @@ from versions before 5.0) that match the expression to be highlighted.
...
-* Treats the document as the whole corpus, and scores individual sentences as
-if they were documents in this corpus, using the  BM25 algorithm
+[float]
+====== Postings
+
+If `index_options` is set to `offsets` in the mapping the `unified` highlighter
+will use this information to highlight documents without re-analyzing the text.
+It re-runs the original query directly on the postings and extracts the matching offsets
+directly from the index limiting the collection to the highlighted documents.
+This mode is faster since it doesn't require to reanalyze the text to be highlighted
```
Comment (nik9000) [inline docs/reference/search/request/highlighting.asciidoc]: I expect it isn't actually faster for very short strings. At least, that was my experience with the experimental highlighter.

File: docs/reference/search/request/highlighting.asciidoc [hunk 9]
```diff
@@ -814,6 +833,8 @@ to
 
 [[phrase-limit]]
 ==== Phrase Limit
+
+WARNING:
```
Comment (nik9000) [inline docs/reference/search/request/highlighting.asciidoc]: Missing "this is only supported by the fast vector highlighter", I think.

**Review by nik9000 — APPROVED** [review]
I left a few minors but LGTM. I guess next is to deprecate the postings highlighter in 5.x, right?

**Comment by jimczi** [comment]
Thanks @clintongormley and @nik9000 
I pushed a commit to address your comments.

> I guess next is to deprecate the postings highlighter in 5.x

We can make this change transparent by allowing `postings` to be set in 6.x as an alias for `unified` (and just add the verification that `postings` is set in the index options) ?

File: core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java (hunk 5 above)
Comment (jimczi) [inline core/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java]: Sorry about that :(

**Comment by nik9000** [comment]
> We can make this change transparent by allowing postings to be set in 6.x as an alias for unified (and just add the verification that postings is set in the index options) ?

I'm afraid of something sneaky like that. I think we're better off deprecating in 5.x so users of the postings highlighter know that they should test with the unified highlighter. Even if it is the same code I'd prefer they know about the change rather than get some kind of surprise on upgrade.

**Comment by javanna** [comment]
Not sure, but I am wondering if we should rather have two separate PRs, one for removing postings which gets replaced by unified (probably already potentially breaking although if it breaks users it's because of bugs? but certainly breaking for people specifying postings as highlighter type), and another one for changing the default highlighter type (more breaking as it affects also people relying on plain or fvh just because they don't have offsets or they have term vectors).

I tend to agree with @nik9000 on not making unified a synonym of postings under the hood.

**Comment by jimczi** [comment]
> I tend to agree with @nik9000 on not making unified a synonym of postings under the hood.

Ok so I'll start with the deprecation in 5.x

> Not sure, but I am wondering if we should rather have two separate PRs, one for removing postings which gets replaced by unified (probably already potentially breaking although if it breaks users it's because of bugs? but certainly breaking for people specifying postings as highlighter type)

I think it affects 5.x only where we have two options:
* Deprecate `postings` but only when the type is forced and redirect automatically to `unified` (instead of `postings`) when the field options for offsets are set.
* Deprecate `postings` but continue to redirect automatically to this highlighter when the field options are set.

I am leaning toward option 2 since the `unified` highlighter also handles phrase/span queries (and the `postings` just ignores the positions).

> and another one for changing the default highlighter type (more breaking as it affects also people relying on plain or fvh just because they don't have offsets or they have term vectors)

If the desired behavior for 6.x is to use the `unified` highlighter by default I don't think we need to split this in two PRs. Doing so would require to change the tests and documentations twice, with a single PR the change is simpler, we don't need to have a replacement for the automatic `postings` mode.

**Comment by jimczi** [comment]
I opened https://github.com/elastic/elasticsearch/pull/25073 for the deprecation in 5.x

**Pull request merged**